lib/libutils/ext/ftrace/ftrace.c

```diff
@@ -62,112 +56,43 @@ static __noprof struct ftrace_buf *get_fbuf(void)
 #endif
 }
 
-#if defined(_CFG_FTRACE_BUF_WHEN_FULL_shift)
-
-/*
- * This API shifts/moves ftrace buffer to create space for new dump
- * in case the buffer size falls short of actual dump.
- */
-static bool __noprof fbuf_make_room(struct ftrace_buf *fbuf, size_t size)
-{
-	char *dst = (char *)fbuf + fbuf->buf_off;
-	const char *src = (char *)fbuf + fbuf->buf_off + size;
-	size_t n = 0;
-
-	fbuf->curr_size -= size;
-
-	for (n = 0; n < fbuf->curr_size; n++)
-		dst[n] = src[n];
-
-	return true;
-}
-
-#elif defined(_CFG_FTRACE_BUF_WHEN_FULL_wrap)
-
-/* Makes room in the trace buffer by discarding the previously recorded data. */
-static bool __noprof fbuf_make_room(struct ftrace_buf *fbuf,
-				    size_t size)
-{
-	if (fbuf->buf_off + size > fbuf->max_size)
-		return false;
-
-	fbuf->curr_size = 0;
-
-	return true;
-}
-
-#elif defined(_CFG_FTRACE_BUF_WHEN_FULL_stop)
-
-static bool __noprof fbuf_make_room(struct ftrace_buf *fbuf __unused,
-				    size_t size __unused)
-{
-	return false;
-}
-
-#else
-#error CFG_FTRACE_BUF_WHEN_FULL value not supported
-#endif
-
-static size_t __noprof to_func_enter_fmt(char *buf, uint32_t ret_idx,
-					 unsigned long pc)
+static void __noprof add_elem(struct ftrace_buf *fbuf, uint8_t level,
+			       uint64_t val)
 {
-	char *str = buf;
-	uint32_t addr_size = 2 * sizeof(unsigned long);
-	uint32_t i = 0;
+	struct __ftrace_elem *elem = NULL;
+	size_t idx = 0;
 
-	for (i = 0; i < (DURATION_MAX_LEN + ret_idx); i++)
-		if (i == (DURATION_MAX_LEN - 2))
-			*str++ = '|';
-		else
-			*str++ = ' ';
+	idx = fbuf->curr_idx;
 
-	*str++ = '0';
-	*str++ = 'x';
+	elem = (struct __ftrace_elem *)((vaddr_t)fbuf + fbuf->buf_off) + idx;
+	elem->stack_depth = level;
+	elem->val = val;
 
-	for (i = 0; i < addr_size; i++)
-		*str++ = hex_str[(pc >> 4 * (addr_size - i - 1)) & 0xf];
+	if (fbuf->curr_size < idx * sizeof(*elem))
+		fbuf->curr_size = idx * sizeof(*elem);
 
-	*str++ = '(';
-	*str++ = ')';
-	*str++ = ' ';
-	*str++ = '{';
-	*str++ = '\n';
-	*str = '\0';
+	idx++;
+	if ((idx + 1) * sizeof(*elem) > fbuf->max_size) {
+		idx = 0;
+		fbuf->curr_size = fbuf->max_size;
```

Comment:
Thinking more about this, I think we should rather have a `overflow` flag instead of `curr_size` since we can very well use `idx` to retrieve the current size. Then we don't need the assignment above too.

Reply:
Agreed.